l1-contracts/src/core/libraries/HeaderLib.sol

```diff
@@ -0,0 +1,154 @@
+  }
+
+  struct PartialStateReference {
+    AppendOnlyTreeSnapshot noteHashTree;
+    AppendOnlyTreeSnapshot nullifierTree;
+    AppendOnlyTreeSnapshot contractTree;
+    AppendOnlyTreeSnapshot publicDataTree;
+  }
+
+  struct StateReference {
+    AppendOnlyTreeSnapshot l1ToL2MessageTree;
+    // Note: Can't use "partial" name here as in yellow paper because it is a reserved solidity keyword
+    PartialStateReference partialStateReference;
+  }
+
+  struct Header {
+    GlobalVariables globalVariables;
+    StateReference stateReference;
+    AppendOnlyTreeSnapshot lastArchive;
+    bytes32 bodyHash;
+  }
+
+  /**
+   * @notice Decodes the header
+   * @param _header - The header calldata
+   * @return The decoded header
+   */
+  function decode(bytes calldata _header) internal pure returns (Header memory) {
+    require(_header.length == 376, "Invalid header length");
+
+    Header memory header;
+
+    header.globalVariables.chainId = uint256(bytes32(_header[:0x20]));
+    header.globalVariables.version = uint256(bytes32(_header[0x20:0x40]));
+    header.globalVariables.blockNumber = uint256(bytes32(_header[0x40:0x60]));
+    header.globalVariables.timestamp = uint256(bytes32(_header[0x60:0x80]));
+    header.stateReference.l1ToL2MessageTree =
+      AppendOnlyTreeSnapshot(bytes32(_header[0x80:0xa0]), uint32(bytes4(_header[0xa0:0xa4])));
+    header.stateReference.partialStateReference.noteHashTree =
+      AppendOnlyTreeSnapshot(bytes32(_header[0xa4:0xc4]), uint32(bytes4(_header[0xc4:0xc8])));
+    header.stateReference.partialStateReference.nullifierTree =
+      AppendOnlyTreeSnapshot(bytes32(_header[0xc8:0xe8]), uint32(bytes4(_header[0xe8:0xec])));
+    header.stateReference.partialStateReference.contractTree =
+      AppendOnlyTreeSnapshot(bytes32(_header[0xec:0x10c]), uint32(bytes4(_header[0x10c:0x110])));
+    header.stateReference.partialStateReference.publicDataTree =
+      AppendOnlyTreeSnapshot(bytes32(_header[0x110:0x130]), uint32(bytes4(_header[0x130:0x134])));
+    header.lastArchive =
+      AppendOnlyTreeSnapshot(bytes32(_header[0x134:0x154]), uint32(bytes4(_header[0x154:0x158])));
+
+    header.bodyHash = bytes32(_header[0x158:0x178]);
+
+    return header;
+  }
+
+  /**
+   * @notice Validates the header
+   * @param _header - The decoded header
+   * @param _version - The expected version
+   * @param _lastBlockTs - The timestamp of the last block
+   * @param _archive - The expected archive root
+   */
+  function validate(Header memory _header, uint256 _version, uint256 _lastBlockTs, bytes32 _archive)
+    internal
+    view
+  {
+    if (block.chainid != _header.globalVariables.chainId) {
+      revert Errors.Rollup__InvalidChainId(_header.globalVariables.chainId, block.chainid);
+    }
+
+    if (_header.globalVariables.version != _version) {
+      revert Errors.Rollup__InvalidVersion(_header.globalVariables.version, _version);
+    }
+
+    // block number already constrained by archive root check
```

Comment:
As commented elsewhere. We need to constrain it in the circuits here instead of the contract.